sql/src/main/java/io/crate/metadata/DefaultTemplateService.java

```diff
@@ -0,0 +1,139 @@
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import io.crate.Constants;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.metadata.IndexTemplateMetaData;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+
+import java.io.IOException;
+
+/**
+ * Service that ensures a "crate_defaults" template exists.
+ *
+ * This template matches all indices and customizes dynamic mapping updates:
+ *
+ *  - dynamically added strings use docValues
+ */
+public class DefaultTemplateService {
+
+    final static String TEMPLATE_NAME = "crate_defaults";
+
+    private final static ESLogger LOGGER = Loggers.getLogger(DefaultTemplateService.class);
+    private final static String DEFAULT_MAPPING_SOURCE = createDefaultMappingSource();
+    private final ClusterService clusterService;
+
+
+    public DefaultTemplateService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    public void createIfNotExists(ClusterState state) {
+        if (clusterService.localNode().isMasterNode() == false) {
+            return;
+        }
+        if (state.getMetaData().getTemplates().containsKey(TEMPLATE_NAME) == false) {
+            createDefaultTemplate();
+        }
+    }
+
+    private void createDefaultTemplate() {
+        clusterService.submitStateUpdateTask("ensure-default-template", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return addDefaultTemplate(currentState);
+            }
+
+            @Override
+            public void onFailure(String source, Throwable cause) {
+                LOGGER.error("Error during ensure-default-template source={}", cause, source);
+            }
+        });
+    }
+
+    @VisibleForTesting
+    static ClusterState addDefaultTemplate(ClusterState currentState) throws IOException {
+        MetaData currentMetaData = currentState.getMetaData();
+        ImmutableOpenMap<String, IndexTemplateMetaData> currentTemplates = currentMetaData.getTemplates();
+        IndexTemplateMetaData defaultTemplate = currentTemplates.get(TEMPLATE_NAME);
+        if (defaultTemplate != null) {
```

Comment:
hm cannot understand how this could happen as only the master node will create it once and the current state is already checked at createIfNotExists...